Tasks/JavaToolInstallerV0/javatoolinstaller.ts

```diff
@@ -94,22 +99,156 @@ function buildFilePath(localPathRoot: string, fileEnding: string, fileNameAndPat
+    }
+    return jdkDirectory;
+}
+
+/**
+ * Get the path to a folder inside the VOLUMES_FOLDER.
+ * Only for macOS.
+ * @param volumes VOLUMES_FOLDER contents before attaching a disk image.
+ */
+function getVolumePath(volumes: Set<string>): string {
+    const newVolumes: string[] = fs.readdirSync(VOLUMES_FOLDER).filter(volume => !volumes.has(volume));
+
+    if (newVolumes.length !== 1) {
+        throw new Error(taskLib.loc('UnsupportedDMGStructure'));
+    }
+    return path.join(VOLUMES_FOLDER, newVolumes[0]);
+}
+
+/**
+ * Get path to a .pkg file.
+ * Only for macOS.
+ * @param volumePath Path to the folder containing a .pkg file.
+ */
+function getPackagePath(volumePath: string): string {
+    const packages: string[] = fs.readdirSync(volumePath).filter(file => file.endsWith('.pkg'));
+
+    if (packages.length === 1) {
+        return path.join(volumePath, packages[0]);
+    } else if (packages.length === 0) {
+        throw new Error(taskLib.loc('NoPKGFile'));
+    } else {
+        throw new Error(taskLib.loc('SeveralPKGFiles'));
+    }
+}
+
+async function installPkg(pkgPath: string): Promise<string> {
+    if (!fs.existsSync(pkgPath)) {
+        throw new Error('PkgPathDoesNotExist');
+    }
+
+    console.log(taskLib.loc('InstallJDK'));
+
+    // Using set because 'includes' array method requires tsconfig option "lib": ["ES2017"]
+    const JDKs: Set<string> = new Set(fs.readdirSync(JDK_FOLDER));
+
+    await runPkgInstaller(pkgPath);
+
+    const newJDKs = fs.readdirSync(JDK_FOLDER).filter(jdkName => !JDKs.has(jdkName));
+
+    if (newJDKs.length !== 1) {
+        throw new Error(taskLib.loc('NewJDKIsNotInstalled'));
+    }
+
+    let jdkDirectory: string = path.join(JDK_FOLDER, newJDKs[0], JDK_HOME_FOLDER);
+    return jdkDirectory;
+}
+
+/**
+ * Run a tool with `sudo` on Linux and macOS.
+ * Precondition: `toolName` executable is in PATH.
+ */
+function sudo(toolName: string): ToolRunner {
+    if (os.platform() === 'win32') {
+        return taskLib.tool(toolName);
     } else {
-        throw new Error(taskLib.loc('UnsupportedFileExtension'));
+        const toolPath = taskLib.which(toolName);
+        return taskLib.tool('sudo').line(toolPath);
+    }
+}
+
+/**
+ * Attach a disk image.
+ * @param sourceFile Path to JDK file.
+ */
+async function attach(sourceFile: string): Promise<void> {
+    console.log(taskLib.loc('AttachDiskImage'));
+    const hdiutil = sudo('hdiutil');
+    hdiutil.line(`attach "${sourceFile}"`);
+    await hdiutil.exec();
+}
+
+/**
+ * Install a .pkg file.
+ * Only for macOS.
+ * @param pkgPath Path to a .pkg file.
+ */
+async function runPkgInstaller(pkgPath: string): Promise<void> {
+    try {
+        const installer = sudo('installer');
+        installer.line(`-package "${pkgPath}" -target /`);
+        await installer.exec();
+    } catch (error) {
+        taskLib.debug('Failed to install pkg file');
```

Comment:
Could you please add `error` to the debug output as well? I think it can be useful.

Reply:
Fair note, thanks! Added it.